### ISSUE TITLE
Improve verbosity of logs

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#4032]
# This is a Feature Proposal

## Description

I found that existing logs that output with `-v` option, doesn't give good overview of internal process.

To have better overview of that, I looked into serverless internals and added for my purpose logs that expose
- Commands that are about to be pursued with their hooks filled and unfilled with actions
- Hook start and end points with _own_ duration it took
- All AWS requests with parameters, so I can see exactly how SLS works with AWS SDK

Here's an example output:

![screen shot 2017-08-03 at 11 45 47](https://user-images.githubusercontent.com/122434/28916177-a1160780-7841-11e7-8e0b-f0d1d2724e6c.png)

![screen shot 2017-08-03 at 11 47 15](https://user-images.githubusercontent.com/122434/28916184-a697ad80-7841-11e7-98b7-df1ff6b1c5c2.png)

I think it'll be nice to have it in Serverless natively (at least I think there's no way to introduce it via plugin), so they show either in _verbose_ mode, or with `SERVERLESS_DEBUG=true`, or with some kind of namespace debugging  (as can be achieved with [debug](https://www.npmjs.com/package/debug)).

I'll be happy to propose a PR, with that, let me know what you think

## Additional Data

* ***Serverless Framework Version you're using***: v1.19.0
* ***Operating System***: macOS v10.12.6

**Comment by horike37** [comment]
@medikoo
Does you want to add the feature to `sls deploy` command?
For sls plugin developer, it might be nice.

**Comment by medikoo** [comment]
@horike37 I think it should be command agnostic. In any case, SLS command is lifecycle of events with attached hooks.

So for any kind of command it'll be great to have the option to see how process is resolved beneath. I think it'll be especially helpful for plugin authors.

Still major question is, when do those logs should be output:
- In _verbose_ mode?
- with `SERVERLESS_DEBUG=true` ?
- Or maybe through some namespace debugging (as can be achieved with [debug](https://www.npmjs.com/package/debug)), but as far as I see, there's no such solution currently in SLS v1

**Comment by HyperBrain** [comment]
@medikoo @horike37 As this feature is only valuable for plugin developers and not for framework users, I do not see it in the core framework. This can and should imo be implemented as plugin, because of the very low percentage of people that need the output.

> at least I think there's no way to introduce it via plugin
My opinion is, that it is possible. Given that a pure developer plugin can use techniques that are more intrusive that production plugins, it should be possible to wrap the hooks and the plugins in pluginManager from the plugin and intercept all functionality there. Even getCommand() or spawn() can be wrapped.
The wrappers should be applied in the plugin's constructor - and the dev-log plugin must be the last one in the service's plugin list.
This is perfectly doable.

**Comment by horike37** [comment]
What about using `SLS_TRACE_HOOK` env vars?
`sls deploy` command has already had `-v` options, so it would be breaking chage.
Then, I prefer using environment variables rather than command line option. Since it would want to enable it for a certain period of time, such as developing a plugin, rather than every time the command runs

**Comment by medikoo** [comment]
> Given that a pure developer plugin can use techniques that are more intrusive than production plugins

@HyperBrain sure, but that's way less reliable, e.g. internal methods may change across minor/patch SLS versions and therefore break plugin which tweaks that.

If it's indeed solvable by plugin then it'll be good to clarify on how to achieve it with guarantee it won't break within given major version scope (?)

**Comment by HyperBrain** [comment]
The location where the plugin would wrap methods and inject into the hooks array it the very central point of pluginManager and the very central point of the core framework. This is the least likely place where even internal changes will happen as it controls the fundamental workflow and all plugins (internal and external ones) depend on its behavior.

Especially the `run()`, `spawn()` and `getCommand()` methods of its interface and the hook and command definitions are fixed as they are used everywhere and from everyone. Even the internal `hooks` and `plugins` arrays are IMO completely stable and fixed.

The more you get away from pluginManager, the higher is the risk of breaking changes - with the highest probability in the packaging and deployment plugin parts of the framework.

**Comment by medikoo** [comment]
@HyperBrain Implementing it in, will require to patch:
- `PluginManager#invoke` (Commands and hooks logging)
- `AwsProvider#request` (AWS SDK calls logging)

So if those methods are not meant to change within scope of major release, then indeed it can be solved via plugin.

@horike37 `SLS_TRACE_HOOK` sounds fine, still what about AWS request, another env var? Aren't we re-implementing functionality of [debug](https://www.npmjs.com/package/debug)) then? :)

**Comment by HyperBrain** [comment]
@medikoo I think so. `AwsProvider#request` is also a central interface method, that is used by core and 3rd party plugins. A change there would break everything.

Additionally to `PluginManager#invoke` I would also patch `PluginManager#spawn` and show a different message because spawn is used to invoke internal plugin lifecycles from plugins. It allows access to `entrypoints` which are not accessible from the CLI. It would be anough to show `Spawned 'XXXXXX'` in that case.

@horike37 @medikoo  I think the core integration solution will introduce too much technical debt and complexity (we're already talking about 2 env vars here).

**Comment by medikoo** [comment]
@HyperBrain Thanks, I'll take that into account.

> @horike37 @medikoo I think the core integration solution will introduce too much technical debt and complexity (we're already talking about 2 env vars here).

I also think it's not best approach. If there's no _namespace_ based logging available within SLS, then there's no lean solution for that.
_As a side note, I saw that while [debug](https://www.npmjs.com/package/debug) was used in SLS v0.5, it's for some reason not present in v1. Do you know why it was decided to not rely on it?_

If we mostly agree that it should be in external plugin, please close, and I will post link to plugin here once it's ready.

**Comment by horike37** [comment]
As @HyperBrain said, we had better be in external plugin if the framework will be complexity by introducing the feature to sls core.

**Comment by pmuens** [comment]
Thanks for creating @medikoo and thanks for commenting @horike37 and @HyperBrain 👍

I agree. Having this as an external would be super nice here (something like an extra `debug` plugin).

> If we mostly agree that it should be in external plugin, please close, and I will post link to plugin here once it's ready.

I'll close this one, but please comment here again if you have any questions or need help. Furthermore feel free to comment once the plugin is done!

Thanks again for opening and kicking off the discussion @medikoo 👍

This issue might be related to https://github.com/serverless/serverless/pull/2821.
